From 2d8d02e87ed3968a8a984cc0c8051a6398de8b64 Mon Sep 17 00:00:00 2001
From: dre698 <dre698@users.noreply.github.com>
Date: Sun, 4 Oct 2026 01:13:59 +0000
Subject: [PATCH 2/3] fs/stat: add ksu_handle_newfstat_ret /
 ksu_handle_fstat64_ret hooks

NadekoSU/ReSukiSU check for the fstat return hooks and stop the build
when they are missing.

* fs/stat: newfstat: ksu_handle_newfstat_ret (required)
* fs/stat: fstat64: ksu_handle_fstat64_ret (32-bit su, optional)
---
 fs/stat.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/fs/stat.c b/fs/stat.c
index 23adf4e..779b484 100644
--- a/fs/stat.c
+++ b/fs/stat.c
@@ -310,6 +310,10 @@ SYSCALL_DEFINE4(newfstatat, int, dfd, const char __user *, filename,
 }
 #endif
 
+#ifdef CONFIG_KSU_MANUAL_HOOK
+extern void ksu_handle_newfstat_ret(unsigned int *fd, struct stat __user **statbuf_ptr);
+#endif
+
 SYSCALL_DEFINE2(newfstat, unsigned int, fd, struct stat __user *, statbuf)
 {
 	struct kstat stat;
@@ -318,6 +322,9 @@ SYSCALL_DEFINE2(newfstat, unsigned int, fd, struct stat __user *, statbuf)
 	if (!error)
 		error = cp_new_stat(&stat, statbuf);
 
+#ifdef CONFIG_KSU_MANUAL_HOOK
+	ksu_handle_newfstat_ret(&fd, &statbuf);
+#endif
 	return error;
 }
 
@@ -428,6 +435,10 @@ SYSCALL_DEFINE2(lstat64, const char __user *, filename,
 	return error;
 }
 
+#ifdef CONFIG_KSU_MANUAL_HOOK
+extern void ksu_handle_fstat64_ret(unsigned long *fd, struct stat64 __user **statbuf_ptr);
+#endif
+
 SYSCALL_DEFINE2(fstat64, unsigned long, fd, struct stat64 __user *, statbuf)
 {
 	struct kstat stat;
@@ -436,6 +447,9 @@ SYSCALL_DEFINE2(fstat64, unsigned long, fd, struct stat64 __user *, statbuf)
 	if (!error)
 		error = cp_new_stat64(&stat, statbuf);
 
+#ifdef CONFIG_KSU_MANUAL_HOOK /* 32-bit su */
+	ksu_handle_fstat64_ret(&fd, &statbuf);
+#endif
 	return error;
 }
 
